File: amic-platform/src/modules/fdd/types/issue.ts
// ── Enums ────────────────────────────────────────────────

export type IssueCategory =
  | "ANOMALY"
  | "DATA_QUALITY"
  | "MAPPING"
  | "CALCULATION"
  | "AI_SUGGESTION";

export type IssueSeverity = "CRITICAL" | "HIGH" | "MEDIUM" | "LOW";

export type IssueStatus =
  | "OPEN"
  | "UNDER_REVIEW"
  | "RESOLVED"
  | "FALSE_POSITIVE"
  | "ACKNOWLEDGED";

// ── Issue ────────────────────────────────────────────────

export interface IssueRead {
  id: string;
  deal_id: string;
  snapshot_id: string | null;
  category: IssueCategory;
  severity: IssueSeverity;
  status: IssueStatus;
  title: string;
  description: string;
  risk_score: string | null; // Decimal as string
  source_type: string | null;
  source_id: string | null;
  source_detail: Record<string, unknown> | null;
  detection_method: string;
  detection_factors: Record<string, unknown>[] | null;
  engine_version: string | null;
  resolution_note: string | null;
  resolved_by: string | null;
  resolved_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface IssueCreate {
  category: IssueCategory;
  severity: IssueSeverity;
  title: string;
  description: string;
  source_type?: string | null;
  source_id?: string | null;
  source_detail?: Record<string, unknown> | null;
  detection_method?: string;
}

export interface IssueUpdate {
  status: IssueStatus;
  resolution_note?: string | null;
  resolved_by?: string | null;
}

// ── Issue List Response ──────────────────────────────────

export interface IssueListResponse {
  items: IssueRead[];
  total: number;
  limit: number;
  offset: number;
}

// ── Issue Summary ────────────────────────────────────────

export interface IssueSummary {
  total: number;
  by_severity: Record<string, number>;
  by_status: Record<string, number>;
  by_category: Record<string, number>;
}

// ── Anomaly Detection ────────────────────────────────────

export interface AnomalyDetectionRequest {
  snapshot_id?: string | null;
  threshold?: string; // Decimal as string
}

export interface AnomalyDetectionResponse {
  detected_count: number;
  issues_created: number;
  engine_version: string;
  threshold_used: string;
}

File: fdd/backend/app/schemas/issue.py
"""이슈 관련 Pydantic 스키마 — Sprint 6."""

import uuid
from datetime import datetime
from decimal import Decimal
from typing import Any

from pydantic import BaseModel, Field

from app.models.issue import IssueCategory, IssueSeverity, IssueStatus

# ── Issue Create/Update ──────────────────────────────────


class IssueCreate(BaseModel):
    """수동 이슈 생성 요청."""

    category: IssueCategory
    severity: IssueSeverity
    title: str = Field(..., min_length=1, max_length=200)
    description: str = Field(..., min_length=1)
    source_type: str | None = None
    source_id: str | None = None
    source_detail: dict[str, Any] | None = None
    detection_method: str = "manual"


class IssueUpdate(BaseModel):
    """이슈 상태 변경 요청."""

    status: IssueStatus
    resolution_note: str | None = None
    resolved_by: str | None = None


# ── Issue Read ────────────────────────────────────────────


class IssueRead(BaseModel):
    """이슈 조회 응답."""

    id: uuid.UUID
    deal_id: uuid.UUID
    snapshot_id: uuid.UUID | None
    category: IssueCategory
    severity: IssueSeverity
    status: IssueStatus
    title: str
    description: str
    risk_score: Decimal | None
    source_type: str | None
    source_id: str | None
    source_detail: dict[str, Any] | None
    detection_method: str
    detection_factors: list[dict] | None
    engine_version: str | None
    resolution_note: str | None
    resolved_by: str | None
    resolved_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


# ── Issue List Response ────────────────────────────────────


class IssueListResponse(BaseModel):
    """이슈 목록 응답 (페이지네이션)."""

    items: list[IssueRead]
    total: int
    limit: int
    offset: int


# ── Issue Summary ──────────────────────────────────────────


class IssueSummary(BaseModel):
    """이슈 요약 통계."""

    total: int
    by_severity: dict[str, int]
    by_status: dict[str, int]
    by_category: dict[str, int]


# ── Anomaly Detection ──────────────────────────────────────


class AnomalyDetectionRequest(BaseModel):
    """이상치 탐지 실행 요청."""

    snapshot_id: uuid.UUID | None = None
    threshold: Decimal = Field(
        default=Decimal("50.0"),
        ge=Decimal("0"),
        le=Decimal("100"),
        description="이상치 판정 임계값 (0-100)",
    )


class AnomalyDetectionResponse(BaseModel):
    """이상치 탐지 실행 응답."""

    detected_count: int = Field(description="탐지된 이상치 수")
    issues_created: int = Field(description="생성된 이슈 수")
    engine_version: str
    threshold_used: Decimal
